### ISSUE TITLE
Questions about implement a new DataSet

### ISSUE THREAD

**Issue by soloman817** [GSI-CS-CO/chart-fx#8]
I would like to implement a DataSet that will efficiently support adding/removing data. The requirement comes from that I want to dynamically plotting some event. These events are not coming in regular time, they can be 1 event per second, or 1000 events per second. For this reason, the buffer size of data points will keep changing, based on time, say, I want to always plot the last 10 minutes data.

Now I checked the default DoubleDataSet implementation, and it uses array: https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java#L27

During adding/removing, a new array will be created and old data will be copied there: https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java#L242

So, I would like to use some ring-buffer for the implementation to make it more efficient. 

Now the question comes, because I saw the interface exposes data as array: https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/DoubleDataSet.java#L125

That means, somewhere else in the framework, perhaps the renderer, they work with array, which means, even if I use ring buffer for the internal implementation of the dataset, the framework still requires making array out of the ring buffer, which turns it again not efficient.

Any ideas or suggestions?

**Comment by RalphSteinhagen** [comment]
The following examples may help/illustrate your use-case:
[RollingBufferSample](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-samples/src/main/java/de/gsi/chart/samples/RollingBufferSample.java)
[RollingBufferSortedTreeSample](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-samples/src/main/java/de/gsi/chart/samples/RollingBufferSortedTreeSample.java)

In particular there are several DataSet implementations that implement (timed) circular buffers (or trees, if data arrives/is added out-of-order), e.g.:
[CircularDoubleErrorDataSet](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/CircularDoubleErrorDataSet.java)
[FifoDoubleErrorDataSet](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/FifoDoubleErrorDataSet.java): 
[LimitedIndexedTreeDataSet](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/LimitedIndexedTreeDataSet.java)

Here 'timed' means that the maximum number of samples, as well as the maximum distance between first and last data points, is limited (ie. points are being dropped if the maximum queue size of 100k points and/or x-range of 10 minutes is exceeded).

You particular use-case of 10 minutes of burst data with 1kHz sampling rates (<->100k data points) should be possible. Please note that the maximum update rate should ideally be throttled to <=25 Hz (ie. adding 80 or more samples per 40 ms update interval, and notify/trigger redraws only at max 25 Hz). You may add more data at any rate, but the internal JavaFX thread won't update at higher rates.

Enjoy!

**Comment by soloman817** [comment]
@RalphSteinhagen great example! checking that

**Comment by soloman817** [comment]
@RalphSteinhagen Thanks for point me out the wonderful example, I've made my first version. Here are some findings:

1. `CircularDoubleErrorDataSet` doesn't support evicting data by time, and for x axis, no positive or negative error.

2. `LimitedIndexedTreeDataSet` has queue size and time duration expiration feature. One thing is, while evicting an expired element, it loops over all element to calculate the range: https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/LimitedIndexedTreeDataSet.java#L135  this looks not very efficient. Another thing is, although it supports x error and y error, but you cannot distinct positive or negative error. I extended a class and made some override to support negative x error, but still need to fix some issue, because the range calculation still considered negative error.

But overall, the system is well designed, very good.
